// @sap/ux-specification/dist/src/sync/v4/generate/generate.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateFreestyleInterfaceV4 = exports.generateListReportInterfaceV4 = exports.generateObjectPageInterfaceV4 = void 0;
const utils_1 = require("../../common/utils");
const _1 = require(".");
const ux_specification_types_1 = require("@sap/ux-specification-types");
const i18n_1 = require("../../../i18n/i18n");
const extensionLogger_1 = require("../../../extensionLogger");
const i18next_1 = __importDefault(require("i18next"));
const importProject_1 = require("../../common/importProject");
/**
 * Generates the app specific schema of an object page out of the generic schema for the same page.
 * Generic types are replaced by information from the app specific annotations.
 *
 * @param genericSchema - generic JSON schema of a list report
 * @param entitySetName - the base entity set of the given page
 * @param annotations - list of all annotation files (file content plus URI as identifier)
 * @param manifest - manifest.json of the app
 * @param {ExtensionLogger} logger - Logger class for logging messages
 * @param {FileData[]} fragments - list of fragment files and their content
 *
 * @returns the app specific JSON schema
 */
function generateObjectPageInterfaceV4(genericSchema, entitySetName, annotations, manifest, logger, fragments) {
    const serviceAVT = (0, utils_1.parseAndMergeAndConvert)(annotations, logger);
    const generateParameters = {
        templateType: ux_specification_types_1.TemplateType.ListReportObjectPageV4,
        fioriElementsVersion: ux_specification_types_1.FioriElementsVersion.v4,
        manifest,
        fragments,
        serviceAVT,
        logger
    };
    const entitySet = (0, importProject_1.getEntitySetByEntitySetName)(entitySetName, serviceAVT);
    if (!entitySet) {
        (0, extensionLogger_1.log)(generateParameters.logger, {
            severity: "error" /* LogSeverity.Error */,
            message: i18next_1.default.t('ENTITYSETNOTFOUND', { entitySet: entitySetName }),
            location: {
                path: ux_specification_types_1.METADATAPATH
            }
        });
    }
    return (0, _1.generateObjectPageSchemaV4)(generateParameters, genericSchema, entitySet);
}
exports.generateObjectPageInterfaceV4 = generateObjectPageInterfaceV4;
/**
 * Generates the app specific schema of a list report out of the generic schema.
 * Generic types are replaced by information from the app specific annotations.
 * @param {TemplateType} templateType - the template type allows distinguishing LR and ALP
 * @param genericSchema - generic JSON schema of a list report
 * @param entitySetName - the base entity set of the given page
 * @param annotations - list of all annotation files (file content plus URI as identifier)
 * @param {ExtensionLogger} logger - Logger class for logging messages
 * @param manifest - manifest.json of the app
 *
 * @returns the app specific JSON schema
 */
function generateListReportInterfaceV4(templateType, genericSchema, entitySetName, annotations, logger, manifest) {
    (0, i18n_1.initI18n)();
    const serviceAVT = (0, utils_1.parseAndMergeAndConvert)(annotations, logger);
    const entitySet = serviceAVT?.entitySets.find((es) => es.name === entitySetName);
    const generateParameters = {
        templateType,
        fioriElementsVersion: ux_specification_types_1.FioriElementsVersion.v4,
        manifest,
        serviceAVT,
        logger
    };
    const templateName = templateType === ux_specification_types_1.TemplateType.ListReportObjectPageV4 ? ux_specification_types_1.v4.FE_TEMPLATE_V4_LIST_REPORT : ux_specification_types_1.v4.FE_TEMPLATE_V4_ALP;
    return (0, _1.generateListReportSchemaV4)(generateParameters, genericSchema, entitySet, templateName);
}
exports.generateListReportInterfaceV4 = generateListReportInterfaceV4;
/**
 * Generates the app specific schema of a freestyle page out of the generic schema.
 * Generic types are replaced by information from the app specific annotations.
 * @param {TemplateType} templateType - the template type allows distinguishing LR and ALP
 * @param genericSchema - generic JSON schema of a list report
 * @param annotations - list of all annotation files (file content plus URI as identifier)
 * @param {ExtensionLogger} logger - Logger class for logging messages
 * @param manifest - manifest.json of the app
 *
 * @returns the app specific JSON schema
 */
function generateFreestyleInterfaceV4(templateType, genericSchema, annotations, manifest, entitySetName, logger) {
    const serviceAVT = (0, utils_1.parseAndMergeAndConvert)(annotations, logger);
    const generateParameters = {
        templateType,
        fioriElementsVersion: ux_specification_types_1.FioriElementsVersion.v4,
        manifest,
        serviceAVT,
        logger
    };
    return (0, _1.generateFreestyleSchemaV4)(generateParameters, genericSchema);
}
exports.generateFreestyleInterfaceV4 = generateFreestyleInterfaceV4;
